### ISSUE TITLE
Apps & Packages Feature Exploration!

### ISSUE THREAD

**Issue by joadoumie** [microsoft/devhome#1984]
# Dev Home Feature Exploration: Apps & Packages
## Please Provide Feedback & Ideas!
Please let us know what you think about this feature by commenting on this issue! We'd love to hear your ideas and feedback! 

## Suggested New Feature: Apps & Packages

This new feature aims to create an all-in-one apps & packages management experience in Dev Home that allows you to manage your apps & package in a single, streamlined interface:

![Apps & Packages Mock Up](https://github.com/microsoft/devhome/assets/98557455/8481d2d8-b805-4408-adcd-08a418bd4319)
![Mock up of Apps & Packages Update Flow](https://github.com/microsoft/devhome/assets/98557455/be369898-33d0-476a-be8b-b444fb3b76d3)

### Scenario

As a developer, I want to be able to easily manage my installed apps & packages in one place so that I can quickly update, pin, uninstall, and view my apps & packages. 

### Additional Details

A more detailed spec is in progress and will be linked here when it is complete.

High level features include the ability to:

- Batch uninstall apps & packages
- Detect and apply updates for packages 
- Easy version control + pinning

**Comment by joadoumie** [comment]
#1983

**Comment by agpt8** [comment]
Here are some thoughts that pop in my head after seeing this:
1.  As you said: "This new feature aims to create an all-in-one apps & packages management experience in Dev Home that allows you to manage your apps & package in a single, streamlined interface:", will this also detect and populate packages and apps that are not dev related? From the screenshots I see Edge Dev listed. If it does not, how is it differentiating between dev and non-dev packages and apps.
2. I install many apps from winget that may not be dev related, will they show up here?
3. Will I be able to manually add a package here? If it is already installed, in theory it should be there in the list but if not, can I add it? If it is not already installed, will this serve as an interface to let me install apps through winget? Also while giving me the ability to view the code/command it will use to do it?
4. I dont see the ability to generate winget configuration files. This is something that Dev Home lacks and this particular GUI looks like a viable option to have that option. Importing, exporting, visualizing, editing winget configurations should be an option in dev home.
5. For all the packages installed, it should also be able to show where they are installed on the system. While at it, for installing winget packages through this interface, having options to install packages through a url, zip, or any other install/download option that winget supports would be very useful!

**Comment by revanmj** [comment]
> Here are some thoughts that pop in my head after seeing this:
> 1.  As you said: "This new feature aims to create an all-in-one apps & packages management experience in Dev Home that allows you to manage your apps & package in a single, streamlined interface:", will this also detect and populate packages and apps that are not dev related? From the screenshots I see Edge Dev listed. If it does not, how is it differentiating between dev and non-dev packages and apps.
> 2. I install many apps from winget that may not be dev related, will they show up here?

To me it looks like it's just a GUI for winget and as far as I know there are no package categories in winget. So filtering just dev related packages would require maintaining some kind of a whitelist in Dev Home.

**Comment by CanePlayz** [comment]
As already mentioned, this seems to be a good WinGet GUI approach. However, I think the extension of being able to directly see the current and newest version of a package as well as its source and having an update button is something that could be quite useful for regular users to have in the Windows settings as well. Dev Home could then offer a list of Apps & Packages that only contains the "dev packages" and similar items.

**Comment by mdanish-kh** [comment]
> 4. I dont see the ability to generate winget configuration files. This is something that Dev Home lacks and this particular GUI looks like a viable option to have that option. Importing, exporting, visualizing, editing winget configurations should be an option in dev home.


FYI: there's now a script now over at [winget-create repo to help you create a winget configuration file](https://github.com/microsoft/devhome/tree/main/sampleConfigurations#create-your-own). Currently it's very limited but it's a start.

> 5. For all the packages installed, it should also be able to show where they are installed on the system and where they came from by hovering over the branches icon under Version and Status column (using winget metadata I guess...?). While at it, for installing winget packages through this interface, having options to install packages through a url, zip, or any other install/download option that winget supports would be very useful!

For portable packages, I imagine this should be possible through winget's metadata since WinGet acts as a logical installer for those packages and dictates where it should be put in the system & Registry. For other installers, it may not be a simple task. I'm thinking this would depend on
- https://github.com/microsoft/winget-cli/issues/2298

**Comment by mdanish-kh** [comment]
### Use case

I don't want to install the package right away, but download its installer for a manual installation later.

### Proposal

I would also like to see support for just downloading the version instead of installing it. WinGet has support for downloading packages via the `winget download` command that can be leveraged here.

---

### Use case

Some packages I know don't behave well in a silent installation. It may default to installation in another location while I perform installations for packages in another drive or custom location.

### Proposal

Some installers expose the ability to give a custom install location via the CLI args and that can be accessed through `InstallerSwitches: Location` field in a winget manifest. DevHome can expose a default install location settings field that users can set and if the installer honors the location switch, the custom location can be passed in. There should also be a way to select an install mode i.e., silent, silentWithProgress or interactive for a package.

**Comment by Trenly** [comment]
I would also like the ability to edit the data of the applications displayed in the list.

Use Case: An application doesn’t write the `DisplayVersion` into registry, making it show up as `Unknown`
Proposal: Have an edit button per app entry where the package properties such as DisplayVersion can be updated. Updating the fields would update the registry values. Requires a user to confirm the changes they make could affect the ability to correlate packages or check for updates

**Comment by thepro-3418** [comment]
You definitely don't need to do that, You can use Widgets for UniGetUI and UniGetUI

**Comment by joadoumie** [comment]
@mdanish-kh - really like this idea.

**Comment by JustMyGithub** [comment]
> ### Use case
> 
> Help users author a WinGet manifest quickly. Additionally, detect when a user has a higher version of an app installed than what is available via a WinGet source. In this case, prompt the user for an opportunity to contribute the newer version to WinGet repo. Related to [microsoft/winget-cli#3327](https://github.com/microsoft/winget-cli/issues/3327)

I don't think it is a good idea that "random user" creates a manifest and provides a url. Too much potential for misuse.
Rather just report the higher local version to the server and if a version has been reported more than n times, a maintainer will look into this

https://github.com/microsoft/winget-cli/issues/3361
